### PR TITLE
Make @cached_property generic over the return type

### LINKED ISSUE #2113: Make @cached_property generic over the value
Currently the following code will infer type `Any` for `Foo().bar`, although it should infer type `str`:

```python
from werkzeug.utils import cached_property

class Foo:
    @cached_property
    def bar(self) -> str:
        return ""

reveal_type(Foo().bar)
```

Making `class cached_property` generic over the value solves this problem and corrects the type of a few fields like `Request.base_url`. I have a PR ready for this that actually removes a few instances of `# type: ignore` in werkzeug code (although it adds two new ones.)

### PR DESCRIPTION
**Pull request by srittau**
- fixes #2113

Checklist:

- [ ] ~~Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.~~
- [ ] ~~Add or update relevant docs, in the docs folder and in code.~~
- [ ] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [ ] Add `.. versionchanged::` entries in any relevant code docs.
- [ ] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

The first two items should not be relevant for this PR, I am unsure about the middle two items.

### CODE REVIEW

**Comment by davidism** [comment]
Please rebase this onto the "2.0.x" branch, since it's a typing change it seems fine for a point release.

File: src/werkzeug/utils.py [hunk 1]
```diff
@@ -75,22 +77,22 @@ def value(self):
...
         obj.__dict__[self.__name__] = value
 
-    def __get__(self, obj: object, type: type = None) -> t.Any:  # type: ignore
+    def __get__(self, obj: object, type: type = None) -> _T:  # type: ignore
         if obj is None:
-            return self
+            return self  # type: ignore
 
-        value = obj.__dict__.get(self.__name__, _missing)
+        value = obj.__dict__.get(self.__name__, _missing)  # type: _T
```
Comment (davidism) [inline src/werkzeug/utils.py:95]: I'd like to avoid type comments, was it necessary here?

Comment (srittau) [reply src/werkzeug/utils.py:95]: Without it, `value` is inferred to be `Any` and I got a mypy warning about not returning `Any`.

**Comment by srittau** [comment]
Closing, will reopen a new PR against the 2.0.x branch.

**Pull request closed without merging**